python/pyarrow/src/arrow/python/api.h

```diff
@@ -17,14 +17,14 @@
 
 #pragma once
 
-#include "arrow_to_pandas.h"
-#include "common.h"
-#include "datetime.h"
-#include "deserialize.h"
-#include "helpers.h"
-#include "inference.h"
-#include "io.h"
-#include "numpy_convert.h"
-#include "numpy_to_arrow.h"
-#include "python_to_arrow.h"
-#include "serialize.h"
+#include "arrow/python/arrow_to_pandas.h"
+#include "arrow/python/common.h"
+#include "arrow/python/datetime.h"
+#include "arrow/python/deserialize.h"
+#include "arrow/python/helpers.h"
+#include "arrow/python/inference.h"
+#include "arrow/python/io.h"
+#include "arrow/python/numpy_convert.h"
+#include "arrow/python/numpy_to_arrow.h"
+#include "arrow/python/python_to_arrow.h"
+#include "arrow/python/serialize.h"
```

Comment:
How about reverting all changes in this pull request except this `arrow/python/` prepending changes?
I think that just prepending `arrow/python/` in `python/pyarrow/src/api.h` works.

Reply:
I can't make this to works as:

1. We would like to get rid of top level inlcudions (`#include "foo.h"`) in all the header files.
2. If I try to prepend `pyarrow/` (which we would like to prepend instead of `arrow/python`) to all inclusions except `api.h` and prepend `arrow/python/` to `python/pyarrow/src/pyarrow/api.h` and tell cmake to install the headers into `python/build/dist/include/pyarrow` or `python/build/dist/include/arrow/python` the complier gets stuck on a point of not finding `arrow/python/*` headers or vice versa.

Sorry if I am confusing things.

Reply:
> 1. We would like to get rid of top level inlcudions (#include "foo.h") in all the header files.

Hmm. I think that https://github.com/apache/arrow/pull/13311#discussion_r925344753 means that "we want to keep backward compatibility for `arrow/python/api.h` if we keep providing `arrow/python/api.h`". @pitrou What was your expectation? Backward compatibility? Or just C++ coding style?

If https://github.com/apache/arrow/pull/13311#discussion_r925344753 is for backward compatibility, we don't need to get rid of top level `#include`s. Because `#include "..."` (not `#include <...>`) searches the current directory as the first candidate:

`/usr/local/include/arrow/python/api.h`:

```cpp
#include "serialize.h" // -> Searches /usr/local/include/arrow/python/serialize.h
```

I think that installing all current `.h` (that use `#include "serialize.h"` style not `#include "arrow/python/serialize.h"` style) to `/usr/local/include/arrow/python/` works (existing `#include <arrow/python/api.h>` works with it) and installing all current `.h` to `/usr/local/include/pyarrow/` works (`#include <pyarrow/api.h>` works with it) too.

(Sorry. My comment https://github.com/apache/arrow/pull/14275#discussion_r984034640 was wrong. We don't need to prepend `arrow/python/` to `#include`s in `api.h`.)

Reply:
> I think that installing all current .h (that use #include "serialize.h" style not #include "arrow/python/serialize.h" style) to /usr/local/include/arrow/python/ works (existing #include <arrow/python/api.h> works with it) and installing all current .h to /usr/local/include/pyarrow/ works (#include <pyarrow/api.h> works with it) too.

I think we are thinking in similar direction. Just to clarify (and I think C++ coding style is also important here, not prepending anything works at the moment) possible ways that the includes structure would work:
- (as is currently) installing all current `.h `(with style `#include "serialize.h"`) to `/usr/local/include/arrow/python/ ` where `#include <arrow/python/api.h>` works
- installing all current `.h `(with style `#include "arrow/python/serialize.h"`) to `/usr/local/include/arrow/python/ ` where  `#include <arrow/python/api.h>` works
- installing all current `.h` (with style `#include "pyarrow/serialize.h"`) to `/usr/local/include/pyarrow/` where `#include <pyarrow/api.h>` works

We think the last structure is clearer and would like to change to that at some point. Antoine proposed the following:

> [@pitrou ] ... add a #warning to some compatibility header files that would include the corresponding new ones (https://en.cppreference.com/w/cpp/preprocessor/error).

So if I understand that clearly, we would for some time use both structures (second and third in the list) and add a deprecation warning to the second?

Reply:
For the last warning topic, we can implement it by just installing the following content to `/usr/local/include/arrow/python/api.h`:

```c
#warning "Use <pyarrow/api.h> instead"
#include <pyarrow/api.h>
```

Reply:
How about moving `python/pyarrow/src/*` to `python/pyarrow/` instead of `python/pyarrow/src/arrow/python/`? If we have `*.h` in `python/pyarrow/`, we can use `#include "pyarrow/XXX.h"` in our `*.h` and `*.cc` with `-Ipython/`.

If it's OK, I can work on it as part of ARROW-17838.

Reply:
Moving PyArrow header files to `python/pyarrow` could be a way to move towards `pyarrow/xxx.h` structure. There would still be a need for a deprecating period I think.

I am sure and aware of the probable change with [ARROW-17838](https://issues.apache.org/jira/browse/ARROW-17838) on this topic. But as for now I would like to include this PR in the 10.0.0 release as I think it doesn't present a big change but a good round-up of the work in https://github.com/apache/arrow/pull/13311.

I need to rebase this PR as the PyArrow C++ test work has been merged. After that I would like to see this work being included for the upcoming release, if agreed.

Reply:
> How about moving `python/pyarrow/src/*` to `python/pyarrow/` instead of `python/pyarrow/src/arrow/python/`? If we have `*.h` in `python/pyarrow/`, we can use `#include "pyarrow/XXX.h"` in our `*.h` and `*.cc` with `-Ipython/`.

@kou I would personally prefer to keep the .cc and .h files in a subdirectory, and not directly in `python/pyarrow`, just for code organization reasons (the pyarrow directory is already a long list of files)